### PR TITLE
Added ability to re-raise a exception.

### PR DESCRIPTION
**Pull request by ollydev**
```pascal
begin
  try
     raise 'Will still runtime error!';
  except
     raise;
  end;
end.
```
Outputs: `Runtime error: "Will still runtime error!" at line 3, column 6`

### CODE REVIEW

File: lpcompiler.pas [hunk 1]
```diff
@@ -2699,7 +2699,7 @@ function TLapeCompiler.ParseExpression(ReturnOn: EParserTokenSet = []; FirstNext
               begin
                 if (Method is TLapeTree_InternalMethod) and
                    TLapeTree_InternalMethod(Method).ForceParam and
-                   (not (Tokenizer.Tok in ReturnOn))
+                   (Tokenizer.Tok <> tk_sym_SemiColon) and (not (Tokenizer.Tok in ReturnOn))
```
Comment (nielsAD) [inline lpcompiler.pas]: The additional check shouldn't be necessary here. Semicolon should be part of `ReturnOn`.

```suggestion
                    (not (Tokenizer.Tok in ReturnOn))
```

File: lptree.pas [hunk 2]
```diff
@@ -4228,10 +4229,58 @@ function TLapeTree_InternalMethod_Goto.Compile(var Offset: Integer): TResVar;
...
+    begin
+      if (Node.Parent is TLapeTree_Try) then
+        with Node.Parent as TLapeTree_Try do
+          if (Node = ExceptBody) then
+            Exit(True);
+
+      Node := Node.Parent;
+    end;
+
+    Exit(False);
```
Comment (nielsAD) [inline lptree.pas]: I'm not really a fan of this style, not really a biggie, but might be better to keep it consistent with other code and use `Result := false`.

```suggestion
    Result := False;
```

File: lptypes.pas [hunk 3]
```diff
@@ -658,6 +658,8 @@ procedure _Insert16(Arr: PUInt16; var Index: Integer);
 procedure _Insert32(Arr: PUInt32; var Index: Integer);
 procedure _Insert64(Arr: PUInt64; var Index: Integer);
 
+operator = (Left, Right: TDocPos): Boolean;
```
Comment (nielsAD) [inline lptypes.pas]: This is not compatible with Delphi, unfortunately.

**Comment by ollydev** [comment]
@nielsAD Perhaps it's best to just to add another internal method named `reraise` ? Trying to get raise to work without parameters with FForceParam  and autoinvoke doesn't seem possible in a tidy way.

File: lpcompiler.pas [hunk 4]
```diff
@@ -2697,10 +2697,7 @@ function TLapeCompiler.ParseExpression(ReturnOn: EParserTokenSet = []; FirstNext
                 _LastNode := _Var
               else
               begin
-                if (Method is TLapeTree_InternalMethod) and
-                   TLapeTree_InternalMethod(Method).ForceParam and
-                   (not (Tokenizer.Tok in ReturnOn))
-                then
+                if (Method is TLapeTree_InternalMethod) and (TLapeTree_InternalMethod(Method).ForceParam and (Tokenizer.Tok in ParserToken_Types + [tk_Identifier])) then
```
Comment (nielsAD) [inline lpcompiler.pas]: Do not change this, and instead add semicolon to `ReturnOn` in `ParseStatement()`

```diff
--- a/lpcompiler.pas
+++ b/lpcompiler.pas
@@ -2868,7 +2868,7 @@ begin
     end
   else if (not (Tokenizer.Tok in ParserToken_BlockEnd)) then
   begin
-    Result := ParseExpression([], False);
+    Result := ParseExpression(ExprEnd, False);
     try
       ParseExpressionEnd(ExprEnd);
     except
```

Can you confirm whether this works?

**Comment by ollydev** [comment]
Thanks @nielsAD I implemented your suggestion and seems to work fine.

```pascal
{$I SRL/OSR.simba}
{$I RSWalker/walker.simba}

begin
  try
    raise 'hellloooo';
  except
    if True then
      raise()
    else
      raise;
  end;
end. 

// Runtime error: "hellloooo" at line 6, column 5
```

File: lpmessages.pas [hunk 5]
```diff
@@ -88,6 +90,7 @@   lpException = class(Exception)
   lpeOutOfTypeRangeLow  = 'Out of type range (value:%d, low:%d)';
   lpeOutOfTypeRangeHigh = 'Out of type range (value:%d, high:%d)';
   lpeOutsideExceptionBlock = 'Can only be used in an except/finally block';
+  lpeOutsideExceptBlock = 'Can only be used in a except block';
```
Comment (nielsAD) [inline lpmessages.pas]: an*

**Comment by nielsAD** [comment]
Looks good if you fix the spelling error!

**Comment by nielsAD** [comment]
👍

**Pull request merged**